// app-3rd/lib/extension/views/SectionList.web.js
import React, { Component } from 'react';
import { View, Text, Dimensions, StyleSheet } from 'react-native';
import { ListView, PullToRefresh } from 'antd-mobile';
import PropTypes from 'prop-types'
var { width, height } = Dimensions.get("window");

/**
 * sections数据的格式
[
	{
		key: groupTime
		data: [
			{
				key: it-time
				value: item
			}
		]

	}
]
 */
export default class SectionList extends Component {
    static propTypes = {
        style: PropTypes.object,
        sections: PropTypes.array, //列表数据
        renderItem: PropTypes.func, // item
        renderSectionHeader: PropTypes.any,
        onRefresh: PropTypes.func, // 
        refreshing: PropTypes.bool, //
        onEndReached: PropTypes.func,//调用onEndReached之前的临界值，单位是像素，react-native 中这个值是一个百分比，需要做一下转换
        ListEmptyComponent: PropTypes.any,//空页面
    }

    static defaultProps = {

    }

    uniqueBaseId = `id-${Date.now()}`;
    separatorCount = 0;
    constructor(props) {
        super(props);

        const getSectionData = (dataBlob, sectionID) => dataBlob[sectionID];
        const getRowData = (dataBlob, sectionID, rowID) => dataBlob[rowID];
        const dataSource = new ListView.DataSource({
            getRowData,
            getSectionHeaderData: getSectionData,
            rowHasChanged: (row1, row2) => row1 !== row2,
            sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
        });
        this.state = {
            dataSource,
        };
    }
    _mergeStyle = () => {
        let newStyle = {
            height: height,
            overflow: 'auto',
        };
        if (this.props.style) {
            newStyle = this.props.style;
            if (!newStyle.height) {
                newStyle.height = height;
            }
            if (!newStyle.overflow) {
                newStyle.overflow = "auto";
            }

        }
        return newStyle;
    }

    componentDidMount() {
        this._genData(this.props.sections);
    }

    componentWillReceiveProps(nextProps) {
        this._genData(nextProps.sections);
    }

    _genData = (sections) => {
        if (!sections)
            return;

        const dataBlobs = {};
        let sectionIDs = [];
        let rowIDs = [];
        for (let i = 0; i < sections.length; i++) {
            let sectionName = sections[i].key;
            let rows = sections[i].data;
            sectionIDs.push(sectionName);
            dataBlobs[sectionName] = sectionName;
            rowIDs[i] = [];
            for (let j = 0; j < rows.length; j++) {
                const rowName = `S${i}, R${j}`;
                rowIDs[i].push(rowName);
                dataBlobs[rowName] = rows[j];
            }
        }
        sectionIDs = [...sectionIDs];
        rowIDs = [...rowIDs];
        this.setState({
            dataSource: this.state.dataSource.cloneWithRowsAndSections(dataBlobs, sectionIDs, rowIDs),
        });
    }

    _renderSeparator = () => {
        this.separatorCount++
        if (this.props.ItemSeparatorComponent) {
            let separatorComponent = null;
            if (typeof this.props.ItemSeparatorComponent === "function") {
                separatorComponent = this.props.ItemSeparatorComponent();
            } else {
                separatorComponent = this.props.ItemSeparatorComponent;
            }
            return (
                <View key={`separator-${this.uniqueBaseId}-${this.separatorCount}`}>
                    {
                        separatorComponent
                    }
                </View>
            )
        }
        return null;
    }

    _renderSectionHeader = (sectionData) => {
        if (this.props.renderSectionHeader) {
            let sectionHeader = null;
            if (typeof this.props.renderSectionHeader === "function") {
                sectionHeader = this.props.renderSectionHeader({ section: { key: sectionData } });
            } else {
                sectionHeader = this.props.renderSectionHeader;
            }
            return (
                <View key={`separator-${this.uniqueBaseId}-${this.separatorCount}`}>
                    {
                        sectionHeader
                    }
                </View>
            )
        }
        return null;
    }

    _renderRow = (rowData, sectionID, rowID, highlightRow) => {
        console.log('====================================');
        console.log(rowData);
        console.log(sectionID);
        console.log(rowID);
        console.log('====================================');
        return (
            <View key={`${this.uniqueBaseId}-${rowID}`}>
                {this.props.renderItem ? this.props.renderItem({ item: rowData, index: rowID }) : null}
            </View>
        );
    }
    /**
     * 无数据时的空白页面
     */
    _renderEmpty = () => {
        let height = this._mergeStyle().height;
        let emptyComponent = null;
        this.props.ListEmptyComponent ? this.props.ListEmptyComponent() : null
        if (this.props.ListEmptyComponent) {
            if (typeof this.props.ListEmptyComponent === "function") {
                emptyComponent = this.props.ListEmptyComponent();
            } else {
                emptyComponent = this.props.ListEmptyComponent;
            }
        }
        return (
            <View style={{ height: height }}>
                {
                    emptyComponent
                }
            </View>
        )
    }
    _pullToRefresh = () => {
        if (this.props.onRefresh) {
            return (
                <PullToRefresh
                    refreshing={this.props.refreshing}
                    onRefresh={this.props.onRefresh}
                />
            )
        }
        return null;
    }

    _renderList = () => {
        return (
            <ListView
                ref={el => this.lv = el}
                initialListSize={20}
                horizontal={this.props.horizontal}
                dataSource={this.state.dataSource}
                renderHeader={this.props.renderHeader ? this.props.renderHeader : () => { }}
                renderFooter={this.props.renderFooter ? this.props.renderFooter : () => { }}
                renderRow={(rowData, sectionID, rowID, highlightRow) => { return this._renderRow(rowData, sectionID, rowID, highlightRow) }}
                renderSeparator={this._renderSeparator}
                onEndReached={this.props.onEndReached ? this.props.onEndReached : () => { console.log("reachEnd") }}
                pageSize={this.props.pageSize ? this.props.pageSize : 15}
                onScroll={() => { console.log('scroll'); }}
                scrollRenderAheadDistance={500}
                pullToRefresh={this._pullToRefresh()}
                renderSectionHeader={sectionData => (
                    <div>{`Task ${sectionData.split(' ')[1]}`}</div>
                )}
                renderSectionHeader={this._renderSectionHeader}
                onEndReachedThreshold={100}
                style={this._mergeStyle()}
            />
        );
    }

    render() {
        if (this.props.sections && this.props.sections.length > 0) {
            return this._renderList();
        }
        return this._renderEmpty();
    }
}

const styles = StyleSheet.create({
    listStyle: {
        height: height,
        overflow: 'auto',
    }
})
