// app-3rd/react-navigation.web.js
import React from 'react'
import { NoMatch, Switch, Route } from 'react-router'
import { View, Text } from 'react-native'

// third lib
import { NavigationActions, StackActions } from './lib/react-navigation/src/react-navigation'

export function createStackNavigator(screens, configure) {
    function createStackView(Screen) {
        class StackView extends React.Component {
            render = () => {
                return (
                    <View>
                        <Text>Header View</Text>
                    </View>)
            }
        }
        return StackView;
    }

    class StackRouter extends React.Component {

        render = () => {
            let initScreen = null;
            if (configure.initialRouteName) {
                let initScreenData = screens[configure.initialRouteName];
                initScreen = <Route path={'/'} exact component={initScreenData.screen} />
            }
            let arrayData = [];
            for (var key in screens) {
                arrayData.push({ key: key, item: screens[key] });
            }
            return (<Switch>
                {initScreen}
                {
                    arrayData.map((item, index) => {
                        return <Route key={item + '-' + index} path={'/' + item.key} exact component={createStackView(item.item.screen)} />
                    })}
                <Route component={NoMatch} />
            </Switch>)
        }
    }

    return StackRouter;
}

export { NavigationActions, StackActions }
