// app-components/index.js

// 组件
import ActionButton from '../../pages/common/components/ActionButton'
import {TextInputNormal,TextInputPassword,TextInputImage} from '../../pages/common/components/TextInputs'
import StatusActionButton from '../../pages/common/components/StatusActionButton'
import LoadingView from '../../pages/common/components/LoadingView'
import NoDataView from '../../pages/common/components/NoDataView'
import ActionModal from '../../pages/common/components/ActionModal'
import BarItems from '../../pages/app/components/BarItems'
import GLDDrawerPaneView from '../../pages/app/components/GLDDrawerPaneView'
import GLDNewMenuView from '../../pages/app/components/GLDNewMenuView'
import ImageChooserView from '../../pages/app/components/ImageChooserView'
import ToOfflineOverLay from '../../pages/app/components/ToOfflineOverLay'
import ToOnlineOverLay from '../../pages/app/components/ToOnlineOverLay'
import ToOnlineDialog from '../../pages/app/components/ToOnlineDialog'
import BimSwitch from '../../pages/app/components/BimSwitch'
import ToCleanDialog from '../../pages/app/components/ToCleanDialog'
import CircleProgressBar from '../../pages/app/components/circlebar/CircleProgressBar'
import {shareboard,share} from '../../pages/common/components/ShareUtil'

export {
    ActionButton,
    StatusActionButton,
    BarItems,
    GLDDrawerPaneView,
    GLDNewMenuView,
    TextInputNormal,
    TextInputPassword, 
    TextInputImage,
    ImageChooserView,
    ActionModal,
    LoadingView,
    NoDataView,
    ToOfflineOverLay,
    ToOnlineOverLay,
    ToOnlineDialog,
    BimSwitch,
    ToCleanDialog,
    CircleProgressBar,
    shareboard,
    share,
}
